src/frameworks/next/index.ts

```diff
@@ -183,79 +223,70 @@ export async function init(setup: any) {
+  for (const [path, route] of Object.entries(prerenderManifest.routes)) {
+    // Skip ISR in the deploy to hosting
+    if (route.initialRevalidateSeconds) {
+      continue;
+    }
+
+    // Skip pages affected by middleware in hosting
+    const matchingMiddleware = middlewareMatchers.find((matcher) =>
+      new RegExp(matcher.regexp).test(path)
     );
-    const prerenderManifest = JSON.parse(prerenderManifestBuffer.toString());
-    for (const path in prerenderManifest.routes) {
-      if (prerenderManifest.routes[path]) {
-        // Skip ISR in the deploy to hosting
-        const { initialRevalidateSeconds } = prerenderManifest.routes[path];
-        if (initialRevalidateSeconds) {
-          continue;
-        }
-
-        // TODO(jamesdaniels) explore oppertunity to simplify this now that we
-        //                    are defaulting cleanURLs to true for frameworks
-
-        // / => index.json => index.html => index.html
-        // /foo => foo.json => foo.html
-        const parts = path
-          .split("/")
-          .slice(1)
-          .filter((it) => !!it);
-        const partsOrIndex = parts.length > 0 ? parts : ["index"];
-        const dataPath = `${join(...partsOrIndex)}.json`;
-        const htmlPath = `${join(...partsOrIndex)}.html`;
-        await mkdir(join(destDir, dirname(htmlPath)), { recursive: true });
-        const pagesHtmlPath = join(sourceDir, distDir, "server", "pages", htmlPath);
-        if (await pathExists(pagesHtmlPath)) {
-          await copyFile(pagesHtmlPath, join(destDir, htmlPath));
-        } else {
-          const appHtmlPath = join(sourceDir, distDir, "server", "app", htmlPath);
-          if (await pathExists(appHtmlPath)) {
-            await copyFile(appHtmlPath, join(destDir, htmlPath));
-          }
-        }
-        const dataRoute = prerenderManifest.routes[path].dataRoute;
-        await mkdir(join(destDir, dirname(dataRoute)), { recursive: true });
-        const pagesDataPath = join(sourceDir, distDir, "server", "pages", dataPath);
-        if (await pathExists(pagesDataPath)) {
-          await copyFile(pagesDataPath, join(destDir, dataRoute));
-        } else {
-          const appDataPath = join(sourceDir, distDir, "server", "app", dataPath);
-          if (await pathExists(appDataPath)) {
-            await copyFile(appDataPath, join(destDir, dataRoute));
-          }
-        }
-      }
+    if (matchingMiddleware) {
+      continue;
+    }
+
+    const isReactServerComponent = route.dataRoute.endsWith(".rsc");
+    const contentDist = join(
+      sourceDir,
+      distDir,
+      "server",
+      isReactServerComponent ? "app" : "pages"
+    );
+
+    const parts = path.split("/").filter((it) => !!it);
+    const partsOrIndex = parts.length > 0 ? parts : ["index"];
+
+    const htmlPath = `${join(...partsOrIndex)}.html`;
+    await mkdir(join(destDir, dirname(htmlPath)), { recursive: true });
+    await copyFile(join(contentDist, htmlPath), join(destDir, htmlPath));
```

Comment:
IMO if we're not able to find files we should error, since that's likely a misunderstanding with the directory structure and we want developers to file issues

Reply:
Especially now that we aren't leaning on next export